### DISCUSSION TITLE
Multi-Client Feature Vote

### DISCUSSION THREAD

**Discussion by hassandraga** [webui-dev/webui#434] [Polls]
WebUI is a GUI library focusing on creating a GUI window using a web browser (_Or WebView_) and letting one user access and use it. This is working great for a desktop GUI software project. However, according to this [discussion](https://github.com/webui-dev/webui/discussions/429) and many old discussions, WebUI cannot be used in web app (_local website_) projects because WebUI, by design, allows one single user (_client_) per window only.

A multi-client feature will be helpful for web apps, but implementing this feature will take time and significant core changes, and it will create those new APIs:

```c
// Allow Multi-Client (NEW)
webui_set_config(multi_client, true);

// Events
void event(webui_event_t* e) {

    // New connection
    if (e->event_type == WEBUI_EVENT_CONNECTED) {

        // Get Client ID (NEW)
        size_t clientID = e->client_id;
    }

    // (NEW APIs)
    webui_run_client(clientID, ...); // Run JavaScript in a specific client
    webui_script_client(clientID, ...); // Run JavaScript in a specific client with response
    webui_show_client(clientID, ...); // Refresh UI of a specific client
    webui_close_client(clientID, ...); // Close UI of a specific client
    webui_navigate_client(clientID, ...); // Navigate in a specific client
    webui_send_raw_client(clientID, ...); // Send raw binary to a specific client

    // (UPDATED APIs)
    webui_script(...); // Return empty error... user should use `webui_script_client()` instead in multi-client mode
    webui_run(...); // Run JavaScript in ALL connected clients !
    webui_show(...); // Refresh UI of ALL connected clients !
    webui_close(...); // Close UI of ALL connected clients !
    webui_navigate(...); // Navigate in ALL connected clients !
    webui_send_raw(...); // Send raw binary to ALL connected clients !

    // Congratulations! Now you have a web app powered by your backend language.
}
```

**Note:** If no multi-client mode is set `webui_set_config(multi_client, ...)`, then everything will stay and work as before.

**Comment by GodGavin** [comment]
of course it should be,and webui's goal should be better than blazor (cause balzor can not write in other languages except c#).But at the beginning webui can offer a interface,the specific implementation can be handed over to other backend languages, since web development and concurrency are strengths of these languages, for example, Go.

**Comment by GodGavin** [comment]
Do we have a discord group?please tell me the name I'd like to chat with people there

**Reply by neroist** [reply]
afaik, not at the moment. should one be made?

**Comment by hassandraga** [comment]
**Implemented:**

- Multi-Client Design (_Heavy Core Change_): https://github.com/webui-dev/webui/commit/88c9996ce6fa9e9a79c3f63814023559b73687d2
- Update Bridge - One-time Token: https://github.com/webui-dev/webui/commit/4caeb82351582d314e948add7c2eba88e0a85547
- New multi-client APIs: https://github.com/webui-dev/webui/commit/92e98e84e26fa52a4d304fbac07c4a5ed3f69893

**Note:**
Build-fix, web app examples and multi-client APIs documentation are coming soon.

**Reply by hassandraga** [reply]
* Cookies System: https://github.com/webui-dev/webui/commit/e8f6a8abdb84c63370a2cab4f65ed91f7e41b8e8
* First multi-client web app example: https://github.com/webui-dev/webui/commit/56f787cf73ab4e6cbb97452bdee5446e3e760de1
